File: packages/worker/src/schemas/link.ts
import * as z from "zod";
import * as zu from "../zod-utils";
import type { TagItem } from "./tags";

/**
 * Schema for link item stored in database.
 */
export const LinkItemSchema = z.strictObject({
  id: z.number(),
  title: z.string(),
  url: z.string(),
  favorite: zu.sqlBool(),
  archive: zu.sqlBool(),
  created_at: zu.unixEpochMs(),
  note: z.string(),
});

/**
 * Type for link item stored in database.
 */
export type LinkItem = z.output<typeof LinkItemSchema>;

/**
 * Type for link item returned from API, including associated tags.
 */
export type LinkItemWithTags = LinkItem & {
  tags: TagItem[];
};

File: mobile/lib/models/link.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'tag.dart';

part 'link.freezed.dart';
part 'link.g.dart';

/// A single link item.
/// This maps to `LinkItemWithTags` type on worker side. However `LinkItem` type is an internal type on worker side
/// and is not referenced directly in the API, so we can change the structure of this class without worrying about breaking API compatibility.
@freezed
abstract class Link with _$Link {
  const factory Link({
    required int id,
    required String title,
    required String url,
    required bool favorite,
    required bool archive,
    required List<Tag> tags,
    @JsonKey(name: 'created_at') required int createdAt,
    required String note,
  }) = _Link;

  factory Link.fromJson(Map<String, dynamic> json) => _$LinkFromJson(json);
}
